ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -524,24 +529,105 @@ KEXT_STATIC int HandleFileOpOperation(
+        // it's possible that we'll miss notifications
+        errno_t toErr = vnode_lookup(newPath, 0 /* flags */, &currentVnode, context);
+        if (0 != toErr)
         {
-            PerfSample hardLinkSample(&perfTracer, PrjFSPerfCounter_FileOp_HardLinkCreated);
+            KextLog_Error("HandleFileOpOperation: vnode_lookup failed, errno %d for path '%s'", toErr, newPath);
+            goto CleanupAndReturn;
+        }
         
-            int kauthResult;
-            int kauthError;
-            if (!ProviderMessaging_TrySendRequestAndWaitForResponse(
-                    root,
-                    MessageType_KtoU_NotifyFileHardLinkCreated,
-                    currentVnode,
-                    vnodeFsidInode,
-                    newPath,
-                    pid,
-                    procname,
-                    &kauthResult,
-                    &kauthError))
+        // Don't expect named stream here as they can't be directly hardlinked or renamed, only the main fork can
+        assert(!vnode_isnamedstream(currentVnode));
+        
+        putCurrentVnode = true;
+        
+        bool isDirectory = (0 != vnode_isdir(currentVnode));
+        
+        VirtualizationRootHandle targetRoot, fromRoot;
+        pid_t pid;
+        bool messageTargetProvider =
+            ShouldHandleFileOpEvent(
+                &perfTracer,
+                context,
+                currentVnode,
+                nullptr, // don't pass path for target provider
+                action,
+                isDirectory,
+                &targetRoot,
+                &pid);
+        bool messageFromProvider =
+            ShouldHandleFileOpEvent(
+                &perfTracer,
+                context,
+                currentVnode,
+                fromPath,
+                action,
+                isDirectory,
+                &fromRoot,
+                &pid);
+        
+        if (!messageTargetProvider && !messageFromProvider)
+        {
+            goto CleanupAndReturn;
+        }
+        
+        FsidInode vnodeFsidInode = Vnode_GetFsidAndInode(currentVnode, context, true /* the inode is used for getting the path in the provider, so use linkid */);
+
+        char procname[MAXCOMLEN + 1];
+        proc_name(pid, procname, MAXCOMLEN + 1);
+
+        {
+            PerfSample sample(&perfTracer, PrjFSPerfCounter_FileOp_HardLinkCreated);
+
+            if (messageTargetProvider)
             {
-                goto CleanupAndReturn;
+                int kauthResult;
+                int kauthError = 0;
+                if (!ProviderMessaging_TrySendRequestAndWaitForResponse(
+                        targetRoot,
+                        MessageType_KtoU_NotifyFileHardLinkCreated,
+                        currentVnode,
+                        vnodeFsidInode,
+                        newPath,
+                        (messageFromProvider && targetRoot == fromRoot) ? fromPath : nullptr, // Only send "from" path if in the same root.
```

Comment:
Just an fyi, things may fail later on when "null" is sent down to the mount.  If your tests are passing I wouldn't worry about this.  I think we can discuss in a later PR if we should switch to "" here or just before sending the message in the lib layer.

Reply:
There's a null check in PrjFSLib, so we never send a `PrjFS_NotificationType_HardLinkCreated` with a `nullptr` to `GVFS.Mount` itself.